python_binding/src/byte_stream.rs

```diff
@@ -0,0 +1,254 @@
+    ///
+    #[text_signature = "($self)"]
+    pub fn flush(&mut self) -> PyResult<()> {
+        info!("Flush all data into Pravega Stream {:?}", self.stream);
+        match self.writer.flush() {
+            Ok(()) => Ok(()),
+            Err(e) => Err(exceptions::PyOSError::new_err(format!(
+                "Error while flushing data into ByteStream {:?}",
+                e
+            ))),
+        }
+    }
+
+    ///
+    /// This is used to read bytes from a Pravega Stream into the specified buffer.
+    /// ```
+    /// import pravega_client;
+    /// manager=pravega_client.StreamManager("tcp://127.0.0.1:9090")
+    /// // lets assume the Pravega scope and stream are already created.
+    /// byte_stream=manager.create_byte_stream("scope", "stream")
+    ///
+    /// byte_stream.write(b"bytes")
+    /// byte_stream.flush();
+    /// buf = bytearray(5)
+    /// byte_stream.readinto(buf)
+    /// buf
+    /// bytearray(b'bytes')
+    /// ```
+    ///
+    #[text_signature = "($self, byteArray)"]
+    #[args(byteArray)]
+    pub fn readinto(&mut self, buf: &PyByteArray) -> PyResult<usize> {
+        trace!("Reading binary data from stream {:?} into buffer", buf);
+        let destination_buffer = unsafe { buf.as_bytes_mut() };
+        match self.reader.read(destination_buffer) {
+            Ok(bytes_read) => Ok(bytes_read),
+            Err(e) => Err(exceptions::PyOSError::new_err(format!(
+                "Error while reading from ByteStream {:?}",
+                e
+            ))),
+        }
+    }
+
+    /// Seek to a position within the stream.
+    /// whence follows the same values as IOBase.seek (https://docs.python.org/3/library/io.html#io.IOBase.seek)
+    /// where:
+    /// ```bash
+    /// 0: from start of the stream
+    /// 1: from current stream position
+    /// 2: from end of the stream
+    /// ```
+    ///
+    pub fn seek(&mut self, position: isize, whence: Option<usize>) -> PyResult<u64> {
+        let pos =
+            match whence.unwrap_or(0) {
+                0 => SeekFrom::Start(position as u64),
+                1 => SeekFrom::Current(position as i64),
+                2 => SeekFrom::End(position as i64),
+                _ => return Err(pyo3::exceptions::PyValueError::new_err(
+                    "whence should be one of 0: seek from start, 1: seek from current, or 2: seek from end",
+                )),
+            };
+        match self.reader.seek(pos) {
+            Ok(new_start_offset) => Ok(new_start_offset),
+            Err(e) => Err(exceptions::PyValueError::new_err(format!(
+                "Error while seeking to offset {:?}",
+                e
+            ))),
+        }
+    }
+
+    ///
+    /// Whether the buffer is seekable; here just for compatibility, it always returns True.
+    ///
+    pub fn seekable(&self) -> PyResult<bool> {
+        Ok(true)
+    }
+
+    ///
+    /// Get the current read position in the Stream.
+    ///
+    pub fn tell(&self) -> PyResult<u64> {
```

Comment:
why do we name it `tell`, is there a naming convention in python?

Reply:
This maps the Python API https://docs.python.org/3/library/io.html#io.IOBase.tell